### PR TITLE
Update cpp examples

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
This PR is an update of the PR #2070 . Thanks @Myles1 for this it!

This PR necessitate the other PR **Enable tests for all the files in the example directory** #2296 to be merged into master before.

Changes on the last commit (modification done after #2070):

- The examples have all been put into files and are now in the `docs/examples` directory. As such, they are all tested (with mode `compile`).
- The section **Specify C++ language in setup.py** have been moved out of the tutorial and have been put at the end of the `.rst` file. This was motivated by the fact that this section wasn't necessary to follow the tutorial, and allow us to make a tutorial that is short and to the point. The reader still have access to the section by going further down.

### CODE REVIEW

File: docs/src/userguide/wrapping_CPlusPlus.rst [hunk 1]
```diff
@@ -187,34 +79,47 @@ class name from Rectangle.h and adjust for Cython syntax, so now it becomes::
...
-            void move(int, int)
-
-Note that the constructor is declared as "except +".  If the C++ code or
+.. literalinclude:: ../../examples/userguide/wrapping_CPlusPlus/Rectangle.pxd
+
+Note that the constructor is declared as "except +". If the C++ code or
 the initial memory allocation raises an exception due to a failure, this
 will let Cython safely raise an appropriate Python exception instead
-(see below).  Without this declaration, C++ exceptions originating from
+(see below). Without this declaration, C++ exceptions originating from
```
Comment (scoder) [inline docs/src/userguide/wrapping_CPlusPlus.rst]: I personally find double spaces between sentences helpful as they provide a clearer separation. Also, at least Emacs can make use of them for jumping from sentence to sentence. I'm not suggesting to blindly add them everywhere, but I also don't see a reason to change them to single spaces when they are there already.

File: docs/src/userguide/wrapping_CPlusPlus.rst [hunk 2]
```diff
@@ -277,19 +162,25 @@ A convenient and safe place to do so is in the `__cinit__` and `__dealloc__` met
...
 section on :ref:`resolving naming conflicts <resolve-conflicts>`.
 
+Compilation and Importing
+=========================
+
+To compile a Cython module, it is necessary to have a :file:`setup.py` file:
+
+.. literalinclude:: ../../examples/userguide/wrapping_CPlusPlus/setup.py
+
+Run ``$python setup.py build_ext --inplace``
```
Comment (scoder) [inline docs/src/userguide/wrapping_CPlusPlus.rst]: Should be `$ python` (to suggest a shell command) and not `$python` (which suggests a shell variable).

File: docs/src/userguide/wrapping_CPlusPlus.rst [hunk 3]
```diff
@@ -277,19 +162,25 @@ A convenient and safe place to do so is in the `__cinit__` and `__dealloc__` met
...
+
+Create :file:`test_import.py`:
```
Comment (scoder) [inline docs/src/userguide/wrapping_CPlusPlus.rst]: First thing that popped up in my mind when I read this line was "why?". The previous paragraph says "it is necessary to have a file ...", so is it also necessary now to have a `test_import.py` file? Probably not. ;)

File: docs/src/userguide/wrapping_CPlusPlus.rst [hunk 4]
```diff
@@ -683,6 +574,73 @@ An additional C++11-only RTTI-related class, ``std::type_index``, is available
...
+
+   setup(ext_modules = cythonize(
+              "rect.pyx",                 # our Cython source
+              sources=["Rectangle.cpp"],  # additional source file(s)
+              language="c++",             # generate C++ code
+         ))
+
+Cython will generate and compile the :file:`rect.cpp` file (from
+:file:`rect.pyx`), then it will compile :file:`Rectangle.cpp`
+(implementation of the ``Rectangle`` class) and link both objects files
```
Comment (scoder) [inline docs/src/userguide/wrapping_CPlusPlus.rst]: "object files"

File: docs/src/userguide/wrapping_CPlusPlus.rst [hunk 5]
```diff
@@ -683,6 +574,73 @@ An additional C++11-only RTTI-related class, ``std::type_index``, is available
...
+       name = "rectangleapp",
+       ext_modules = cythonize('*.pyx'),
+   )
+
+And in the .pyx source file, write this into the first comment block, before
+any source code, to compile it in C++ mode and link it statically against the
+:file:`Rectangle.cpp` code file::
+
+   # distutils: language = c++
+   # distutils: sources = Rectangle.cpp
```
Comment (scoder) [inline docs/src/userguide/wrapping_CPlusPlus.rst]: I think it's worth repeating here that the file path is relative to the working directory of the distutils run (which is usually the project root where the `setup.py` resides).

Comment (gabrieldemarmiesse) [reply docs/src/userguide/wrapping_CPlusPlus.rst]: I added a note below.

File: docs/src/userguide/wrapping_CPlusPlus.rst [hunk 6]
```diff
@@ -174,13 +174,19 @@ To compile a Cython module, it is necessary to have a :file:`setup.py` file:
...
+    >>> import rect
+    >>> x0, y0, x1, y1 = 1, 2, 3, 4
+    >>> rect_obj = rect.PyRectangle(x0, y0, x1, y1)
+    >>> print(dir(rect_obj))
+    ['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__',
+     '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__',
+     '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__',
+     '__setstate__', '__sizeof__', '__str__', '__subclasshook__', 'get_area', 'get_size', 'move']
 
-You can now test your wrapper with ``$python test_import.py``.
```
Comment (scoder) [inline docs/src/userguide/wrapping_CPlusPlus.rst]: Ah, yes, that is *much* better.

**Pull request merged**